dbux-code/src/dataFlowView/DataFlowNodeProvider.js

```diff
@@ -0,0 +1,64 @@
+import EmptyArray from '@dbux/common/src/util/EmptyArray';
+import allApplications from '@dbux/data/src/applications/allApplications';
+import traceSelection from '@dbux/data/src/traceSelection';
+import BaseTreeViewNodeProvider from '../codeUtil/BaseTreeViewNodeProvider';
+import TraceNode from '../traceDetailsView/nodes/TraceNode';
+import DataFlowViewModeType from './DataFlowViewModeType';
+import EmptyNode from './EmptyNode';
+import EmptyDataNode from './EmptyDataNode';
+
+/** @typedef {import('@dbux/common/src/core/data/Trace').default} Trace */
+
+export default class DataFlowNodeProvider extends BaseTreeViewNodeProvider {
+  constructor() {
+    super('dbuxDataFlowView');
+  }
+
+  // ###########################################################################
+  // tree building
+  // ###########################################################################
+
+  buildRoots() {
+    const roots = [];
+
+    if (traceSelection.selected) {
+      const trace = traceSelection.selected;
+
+      roots.push(
+        ...this.buildDataNodes(trace)
+      );
+    }
+
+    if (!roots.length) {
+      roots.push(EmptyNode.instance);
+    }
+
+    return roots;
+  }
+
+  /**
+   * @param {Trace} trace
+   */
+  buildDataNodes(trace) {
+    const { applicationId, traceId } = trace;
+    const dp = allApplications.getById(applicationId).dataProvider;
+    const dataNode = dp.indexes.dataNodes.byTrace.getFirst(traceId);
+    if (!dataNode) {
+      return [EmptyDataNode.instance];
+    }
+    
+    const { accessId, valueId } = dataNode;
+    let childDataNodes;
```

Comment:
I think, the `dataNodes` do not have a "child" relationship with `trace` or with `dataNode`. Maybe we can rename to `dataNodes`. It's more ambiguous, but at least less incorrect hah

Reply:
Fix in next PR